retrieval/discovery/kubernetes.go

```diff
@@ -0,0 +1,582 @@
+			panic(fmt.Errorf("discovery.kubernetes.Sources: %s", err))
+		}
+		if res.StatusCode != http.StatusOK {
+			panic(fmt.Errorf("discovery.kubernetes.Sources: %s", res.Body))
+		}
+		var services kubernetes.ServiceList
+		if err := json.NewDecoder(res.Body).Decode(&services); err != nil {
+			panic(fmt.Errorf("discovery.kubernetes.Sources: %s", err))
+		}
+		if err != nil {
+			// If we can't list nodes then we can't watch them. Assume this is a misconfiguration
+			// & panic rather than continue in a bad state
+			panic(fmt.Errorf("discovery.kubernetes.Sources: %s", err))
+		}
+		kd.servicesMu.Lock()
+		defer kd.servicesMu.Unlock()
+
+		kd.servicesResourceVersion = services.ResourceVersion
+		for idx, service := range services.Items {
+			if kd.checkScrape(&service) {
+				sourceNames = append(sourceNames, kubernetesSourceServicePrefix+":"+service.ObjectMeta.Namespace+"/"+service.ObjectMeta.Name)
+				namespace, ok := kd.services[service.ObjectMeta.Namespace]
+				if !ok {
+					namespace = make(map[string]*kubernetes.Service)
+					kd.services[service.ObjectMeta.Namespace] = namespace
+				}
+				namespace[service.ObjectMeta.Name] = &services.Items[idx]
+			}
+		}
+	}
+
+	return sourceNames
+}
+
+func (kd *KubernetesDiscovery) checkScrape(s *kubernetes.Service) bool {
+	if scrape, ok := s.ObjectMeta.Annotations[kd.conf.ScrapeKey]; ok {
+		shouldScrape, err := strconv.ParseBool(scrape)
+		if err != nil {
+			log.Errorf("Invalid value (%s) for %s annotation on service %s/%s: %s", scrape, kd.conf.ScrapeKey, s.ObjectMeta.Namespace, s.ObjectMeta.Name, err)
+			return false
+		}
+		if shouldScrape {
+			return shouldScrape
+		}
+	}
+	return false
+}
+
+// Run implements the TargetProvider interface.
+func (kd *KubernetesDiscovery) Run(ch chan<- *config.TargetGroup) {
+	defer close(ch)
+
+	kd.updateNodesTargetGroup(ch)
+
+	for _, ns := range kd.services {
+		for _, service := range ns {
+			kd.addService(service, ch)
+		}
+	}
+
+	retryDelay := time.Duration(kd.conf.RetryDelay)
+
+	update := make(chan interface{}, 10)
+	go kd.watchNodes(update, retryDelay)
+
+	if kd.conf.DiscoverServices {
+		go kd.watchServices(update, retryDelay)
+		go kd.watchServiceEndpoints(update, retryDelay)
+	}
+
+	for {
+		select {
+		case <-kd.runDone:
+			close(update)
+			close(kd.runDone)
+			return
+		case event := <-update:
+			switch obj := event.(type) {
+			case *kubernetes.NodeEvent:
+				kd.updateNode(obj.Node, obj.EventType)
+				kd.updateNodesTargetGroup(ch)
+			case *kubernetes.ServiceEvent:
+				kd.updateService(obj.Service, obj.EventType, ch)
+			case *kubernetes.EndpointsEvent:
+				kd.updateServiceEndpoints(obj.Endpoints, obj.EventType, ch)
+			}
+		}
+	}
+}
+
+// Stop implements the TargetProvider interface.
+func (kd *KubernetesDiscovery) Stop() {
+	log.Debugf("Stopping Kubernetes discovery for %s", kd.conf.Server)
+
+	// The lock prevents Run from terminating while the watchers attempt
+	// to send on their channels.
```

Comment:
This can hopefully be removed after the pipelining improvements in #977.
Don't worry about the changes over there for now - I'll adjust as necessary later.